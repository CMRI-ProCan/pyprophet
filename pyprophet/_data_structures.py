import pandas as pd
import numpy as np
import random

try:
    profile
except:
    profile = lambda x: x

profile = lambda x: x
from numba import jit, i8, f8



@profile
def prepare_data_table(table, tg_id_name = "transition_group_id",
                              decoy_name = "decoy",
                              main_score_name = None,
                              **extra_args_to_dev_null
                              ):

    column_names = table.columns.values

    # check if given column names appear in table:
    assert tg_id_name in column_names, "colum %s not in table" % tg_id_name
    assert decoy_name in column_names, "colum %s not in table" % decoy_name
    if main_score_name is not None:
        assert main_score_name in column_names,\
                                      "colum %s not in table" % main_score_name

    # if no main_score_name provided, look for unique column with name
    # starting with "main_":
    else:
        main_columns = [c for c in column_names if c.startswith("main_")]
        if not main_columns:
            raise Exception("no column with main_* in table")
        if len(main_columns)>1:
            raise Exception("multiple columns with name main_* in table")
        main_score_name = main_columns[0]

    # get all other score columns, name beginning with "var_"
    var_columns = [c for c in column_names if c.startswith("var_")]
    if not var_columns:
        raise Exception("no column with name var_* in table")

    # collect needed data:
    column_names = """tg_id tg_num_id is_decoy is_top_peak is_train
                      main_score""".split()
    N = len(table)
    empty_col = [0] * N
    empty_none_col = [None] * N

    tg_ids = table[tg_id_name]

    tg_map = dict()
    for i, tg_id in enumerate(tg_ids.unique()):
        tg_map[tg_id] = i
    tg_num_ids = [ tg_map[tg_id] for tg_id in tg_ids]

    data = dict(tg_id = tg_ids.values,
                tg_num_id = tg_num_ids,
                is_decoy = table[decoy_name].values.astype(bool),
                is_top_peak = empty_col,
                is_train = empty_none_col,
                main_score = table[main_score_name].values,
                )

    for i, v in enumerate(var_columns):
        col_name = "var_%d" % i
        data[col_name] = table[v].values
        column_names.append(col_name)

    data["classifier_score"] = empty_col
    column_names.append("classifier_score")

    # build data frame:
    df = pd.DataFrame(data, columns=column_names)

    # for each transition group: enumerate peaks in this group, and
    # add peak_rank where increasing rank corresponds to decreasing main
    # score. peak_rank == 0 is peak with max main score
    return df

@profile
def read_csv(path, sep=None, **extra_args_to_dev_null):
    table = pd.read_csv(path, sep)
    return table

class Experiment(object):

    @staticmethod
    def from_csv(path, **kw):
        raw = read_csv(path, **kw)
        return Experiment.from_data_frame(raw, **kw)

    @staticmethod
    def from_data_frame(df, **kw):
        return Experiment(prepare_data_table(df), **kw)

    @profile
    def __init__(self, df, **kw_to_dev_null):

        var_columns = [c for c in df.columns.values if c.startswith("var_")]
        self.score_columns = var_columns + [ "main_score"]

        # LATER OBSOLETE !
        df.sort("tg_id", ascending=True, inplace=True)

        flags = find_top_ranked(df.tg_num_id.values, df.main_score.values)
        df.is_top_peak = flags
        self.df = df


    @profile
    def split(self, frac):
        decoy_ids = self.df[self.df.is_decoy==True].tg_num_id.unique()
        target_ids = self.df[self.df.is_decoy==False].tg_num_id.unique()

        #random.shuffle(decoy_ids)
        #random.shuffle(target_ids)

        decoy_ids = decoy_ids[:int(len(decoy_ids) * frac)+1]
        target_ids = target_ids[:int(len(target_ids) * frac)+1]

        learn_ids = np.concatenate((decoy_ids,target_ids))
        ix_learn = self.df.tg_num_id.isin(learn_ids)

        self.df.is_train[ix_learn] = True
        self.df.is_train[~ix_learn] = False
