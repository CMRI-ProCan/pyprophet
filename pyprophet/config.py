# encoding: latin-1

# openblas + multiprocessing crashes for OPENBLAS_NUM_THREADS > 1 !!!
import os
os.putenv("OPENBLAS_NUM_THREADS", "1")

import multiprocessing
import random
import sys

import numpy as np
import pandas as pd

# Parameter transformation functions
def transform_pi0_lambda(ctx, param, value):
  if value[1] == 0 and value[2] == 0:
      pi0_lambda = value[0]
  elif 0 <= value[0] < 1 and value[0] <= value[1] <= 1 and 0 < value[2] < 1:
      pi0_lambda = np.arange(value[0], value[1], value[2])
  else:
      sys.exit('Error: Wrong input values for pi0_lambda. pi0_lambda must be within [0,1).')
  return(pi0_lambda)

def transform_threads(ctx, param, value):
    if value == -1:
        value = multiprocessing.cpu_count()
    return(value)

def transform_random_seed(ctx, param, value):
    if value is None:
        value = np.random.randint(0, sys.maxsize)
    return(value)

def transform_subsample_ratio(ctx, param, value):
    if value < 0 or value > 1:
      sys.exit('Error: Wrong input values for subsample_ratio. subsample_ratio must be within [0,1].')
    return(value)

def set_parameters(xeval_fraction, xeval_iterations, initial_fdr, iteration_fdr, ss_iterations, group_id, parametric, pfdr, pi0_lambda, pi0_method, pi0_smooth_df, pi0_smooth_log_pi0, lfdr_truncate, lfdr_monotone, lfdr_transformation, lfdr_adj, lfdr_eps, tric_chromprob, threads, test, random_seed):

    options = dict()

    # Semi-supervised learning
    options['xeval.fraction'] = xeval_fraction
    options['xeval.num_iter'] = xeval_iterations
    options['semi_supervised_learner.initial_fdr'] = initial_fdr
    options['semi_supervised_learner.initial_lambda'] = pi0_lambda
    options['semi_supervised_learner.iteration_fdr'] = iteration_fdr
    options['semi_supervised_learner.iteration_lambda'] = pi0_lambda
    options['semi_supervised_learner.num_iter'] = ss_iterations

    # Statistics
    options['group_id'] = group_id
    options['final_statistics.emp_p'] = parametric
    options['final_statistics.pfdr'] = pfdr
    options['final_statistics.lambda'] = pi0_lambda
    options['final_statistics.pi0_method'] = pi0_method
    options['final_statistics.pi0_smooth_df'] = pi0_smooth_df
    options['final_statistics.pi0_smooth_log_pi0'] = pi0_smooth_log_pi0
    options['final_statistics.lfdr_trunc'] = lfdr_truncate
    options['final_statistics.lfdr_monotone'] = lfdr_monotone
    options['final_statistics.lfdr_transf'] = lfdr_transformation
    options['final_statistics.lfdr_adj'] = lfdr_adj
    options['final_statistics.lfdr_eps'] = lfdr_eps

    # TRIC
    options['tric_chromprob'] = tric_chromprob

    # Processing
    options['num_processes'] = threads
    options['is_test'] = test
    options['random_seed'] = random_seed

    CONFIG.update(options)

def set_pandas_print_options():
    # w, h = pd.util.terminal.get_terminal_size()

    # set output options for regression tests on a wide terminal
    pd.set_option('display.width', 100)
    # reduce precision to avoid to sensitive tests because of roundings:
    pd.set_option('display.precision', 6)


class _ConfigHolder(object):

    def __init__(self):
        self.config = {'num_processes': 1}

    def update(self, dd):
        self.config.update(dd)

    def get(self, name, default=None):
        if default is not None:
            raise RuntimeError("default value not allowed")
        value = self.config.get(name, default)
        return self._translate(name, value)

    def _translate(self, name, value):
        if name == "num_processes" and value == -1:
            value = multiprocessing.cpu_count()
        return value

    def __getitem__(self, name):
        return self.config[name]

    def __setitem__(self, name, value):
        self.config[name] = value

CONFIG = _ConfigHolder()
