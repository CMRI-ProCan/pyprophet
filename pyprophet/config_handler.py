import json

from util import bunchify

default_config = dict(
        num_cross_eval = 5,
        training_proportion = 0.5,
        storey_lambda = 0.4,
        init_run = dict(
            select_positives_with_fdr = 0.15,
        ),
        iter_run = dict(
            select_positives_with_fdr = 0.10,
            )
        )

default_config = bunchify(default_config)

def write_config(cf, path):
    with open(path, "w") as fp:
        json.dump(cf, fp, indent=4)

def read_config(path):
    with open(path, "r") as fp:
        return bunchify(json.load(fp))

def update(cf, key, value):
    if "." in key:
        key_path, final_key = key.rsplit(".", 1)
        for sub_key in key_path.split("."):
            cf = cf.get(sub_key)
            if cf is None:
                raise KeyError("invalid key %s" % key)
        if final_key not in cf:
            raise KeyError("invalid key %s" % key)
    else:
        final_key = key
    cf[final_key] = value

def iter_key_pathes(cf):
    assert isinstance(cf, dict)
    for k, v in cf.items():
        if isinstance(v, dict):
            for p in iter_key_pathes(v):
                yield "%s.%s" % (k, p)
        else:
            yield k

def get_key_pathes(cf):
    return list(iter_key_pathes(cf))
