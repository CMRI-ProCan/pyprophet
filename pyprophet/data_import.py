import pandas as pd

try:
    profile
except:
    profile = lambda x: x

def prepare_data_table(table, tg_id_name = "transition_group_id",
                              decoy_name = "decoy",
                              main_score_name = None,
                              **extra_args_to_dev_null
                              ):

    column_names = table.columns.values

    # check if given column names appear in table:
    assert tg_id_name in column_names, "colum %s not in table" % tg_id_name
    assert decoy_name in column_names, "colum %s not in table" % decoy_name
    if main_score_name is not None:
        assert main_score_name in column_names,\
                                      "colum %s not in table" % main_score_name

    # if no main_score_name provided, look for unique column with name
    # starting with "main_":
    else:
        main_columns = [c for c in column_names if c.startswith("main_")]
        if not main_columns:
            raise Exception("no column with main_* in table")
        if len(main_columns)>1:
            raise Exception("multiple columns with name main_* in table")
        main_score_name = main_columns[0]

    # get all other score columns, name beginning with "var_"
    var_columns = [c for c in column_names if c.startswith("var_")]
    if not var_columns:
        raise Exception("no column with name var_* in table")

    # collect needed data:
    column_names = """tg_id is_decoy main_score peak_rank
                        is_train global_peak_id tg_peak_id""".split()
    N = len(table)
    empty_col = [0] * N
    empty_none_col = [None] * N

    data = dict(tg_id = table[tg_id_name].values,
                is_decoy = table[decoy_name].values.astype(bool),
                main_score = table[main_score_name].values,
                peak_rank = empty_col,
                is_train = empty_none_col,
                global_peak_id = empty_none_col,
                tg_peak_id = empty_none_col,
                )
    for i, v in enumerate(var_columns):
        col_name = "var_%d" % i
        col_name = v
        data[col_name] = table[v].values
        column_names.append(col_name)

    # build data frame:
    df_orig = pd.DataFrame(data, columns=column_names)
    df_orig["global_peak_id"] = df_orig.index  # for reference with full table

    df_orig.sort(("tg_id", "main_score"), ascending=(True, False),
            inplace=True)

    # for each transition group: enumerate peaks in this group, and
    # add peak_rank where increasing rank corresponds to decreasing main
    # score. peak_rank == 0 is peak with max main score
    return df_orig
