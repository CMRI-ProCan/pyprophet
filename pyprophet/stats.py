import numpy
from   util import Bunch

def secure_divide(denom, nom, replace_zero_division_by=0):
    # avoids division by zero for numpy array
    mask = (nom == 0)
    # prepare division without 'division by zero'
    nom[mask] = 1.0
    result = denom / nom
    # replace 'division by zero' places by default:
    result[mask] = replace_zero_division_by
    # restore input array:
    nom[mask] = 0
    return result


def fdr_statistics(p_values, cut_off, lambda_ = 0.5):

    # cast p-values to column vector without nan values

    if isinstance(p_values, (list, tuple)):
        p_values = numpy.array(p_values)
    p_values = p_values[~numpy.isnan(p_values)]
    if p_values.ndim == 1:
        p_values = p_values[:, None]

    assert p_values.shape[-1] == 1 and p_values.ndim == 2

    # cast cut off values to row vector
    cut_off_is_float = isinstance(cut_off, float)
    if cut_off_is_float:
        cut_off = numpy.array([cut_off])
    elif isinstance(cut_off, (list, tuple)):
        cut_off = numpy.array(cut_off)

    if cut_off.ndim == 1:
        cut_off = cut_off[None, :]

    assert cut_off.shape[0] == 1 and cut_off.ndim == 2

    # now broadcasting works eg p_values <= cut_off !!!!!!!!!!!!!!!

    # estimated nulls
    num_null = 1.0 / (1.0-lambda_) * (p_values >= lambda_).sum()
    # fdr
    num_positive = (p_values <= cut_off).sum(axis=0)
    num_null_positive = num_null * cut_off
    fdr = secure_divide(num_null_positive, num_positive, 0).flatten()
    # sensitivity
    num_total = len(p_values)
    num_alternative = num_total - num_null
    num_alternative_positive = num_positive - num_null_positive
    if num_alternative == 0:
        sens = numpy.zeros_like(num_alternative_positive).flatten()
    else:
        sens = num_alternative_positive.flatten() / num_alternative
    # more stat
    num_negative = num_total - num_positive
    num_null_negative= num_null - num_null_positive
    num_alternative_negative = num_negative - num_null_negative

    if cut_off_is_float:
        conv = lambda a: float(a)
    else:
        conv = lambda a: a

    return Bunch(num_total = conv(num_total),
                 num_null  = conv(num_null),
                 num_alternative = conv(num_alternative),
                 num_positive = conv(num_positive),
                 num_negative = conv(num_negative),
                 num_null_positive = conv(num_null_positive),
                 num_alternative_positive = conv(num_alternative_positive),
                 num_null_negative = conv(num_null_negative),
                 num_alternative_negative = conv(num_alternative_negative),
                 fdr = conv(fdr),
                 sens = conv(sens))
