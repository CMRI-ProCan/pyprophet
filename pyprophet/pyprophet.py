# encoding: latin-1

# openblas + multiprocessing crashes for OPENBLAS_NUM_THREADS > 1 !!!
import os
os.putenv("OPENBLAS_NUM_THREADS", "1")

try:
	profile  # ignore
except:
	profile = lambda x: x

import pandas as pd
import numpy as np

from stats import (lookup_s_and_q_values_from_error_table, calculate_final_statistics,
				   mean_and_std_dev, final_err_table, summary_err_table)
from config import CONFIG

from data_handling import (prepare_data_table, Experiment)
from classifiers import (LDALearner, ConsensusPredictor)
from semi_supervised import (AbstractSemiSupervisedTeacher, StandardSemiSupervisedTeacher)

from error_table import FlexibleErrorTable
from null_model import *
from fdr_calc import *  

import multiprocessing

import logging

import time


def unwrap_self_for_multiprocessing((inst, method_name, args),):
	""" You can not call methods with multiprocessing, but free functions,
		If you want to call  inst.method(arg0, arg1),

			unwrap_self_for_multiprocessing(inst, "method", (arg0, arg1))

		does the trick.
	"""
	return getattr(inst, method_name)(*args)


class HolyGostQuery(object):

	""" HolyGhostQuery assembles the unsupervised methods.
		See below how PyProphet parameterises this class.
	"""

	def __init__(self, semi_supervised_teacher):
		assert isinstance(semi_supervised_teacher,
						  AbstractSemiSupervisedTeacher)
		self.semi_supervised_teacher = semi_supervised_teacher

	@profile
	def process_csv(self, path, delim=",", loaded_scorer=None):
		start_at = time.time()

		logging.info("read %s" % path)

		table = pd.read_csv(path, delim, na_values=["NA", "NaN", "infinite"])

		if loaded_scorer is not None:
			logging.info("apply scorer to  %s" % path)
			result_tables, clf_scores, data_for_persistence = self.apply_loaded_scorer(table, loaded_scorer)
			data_for_persistence = None
		else:
			logging.info("learn and apply scorer to %s" % path)
			result_tables, clf_scores, data_for_persistence = self.tutor_and_apply_classifier(table)
		logging.info("processing %s finished" % path)

		needed = time.time() - start_at
		hours = int(needed / 3600)
		needed -= hours * 3600

		minutes = int(needed / 60)
		needed -= minutes * 60

		logging.info("time needed: %02d:%02d:%.1f" % (hours, minutes, needed))
		return result_tables, clf_scores, data_for_persistence


	@profile
	def tutor_and_apply_classifier(self, table):

		prepared_table, score_columns = prepare_data_table(table)

		experiment = Experiment(prepared_table)

		is_test = CONFIG.get("is_test", False)

		if is_test:  # for reliable results
			experiment.df.sort("tg_id", ascending=True, inplace=True)

		experiment.log_summary()

		all_test_target_scores = []
		all_test_decoy_scores = []
		clfs = []
		
		train_frac	 = CONFIG.get("train.fraction")
		is_test	 = CONFIG.get("is_test", False)
		neval		 = CONFIG.get("xeval.num_iter")
		fraction = CONFIG.get("xeval.fraction")
		teacher	 = self.semi_supervised_teacher
		num_processes = CONFIG.get("num_processes")
		
		# reserve part of experiment for testing and FDR calc.
		experiment.split_train_test(train_frac, is_test)
		train_exp = experiment
		test_exp = None
		if train_frac < 0.99:
			train_exp, test_exp = experiment.get_train_and_test_peaks()
		
		xval_type = CONFIG.get("xval.type")
		if xval_type == "split":
			train_exp.set_xval_sets(neval, is_test)
			xval_sets = xval_sets(neval, int(fraction * neval + 0.5))
				
		logging.info("start %d cross evals using %d processes" % (neval, num_processes))
		if num_processes == 1:
			for k in range(neval):
				if xval_type == "split":
					train_exp.train_on_xval_sets(xval_sets[k])
				else:
					train_exp.split_train_test(fraction, is_test)
				(ttt_scores, ttd_scores, clf) = teacher.tutor_randomized(train_exp)
				all_test_target_scores.extend(ttt_scores)
				all_test_decoy_scores.extend(ttd_scores)
				clfs.append(clf)
		else:
			pool = multiprocessing.Pool(processes=num_processes)
			while neval:
				remaining = max(0, neval - num_processes)
				todo = neval - remaining
				neval -= todo
				args = ((teacher, "tutor_randomized", (train_exp, )), ) * todo
				res = pool.map(unwrap_self_for_multiprocessing, args)
				top_test_target_scores = [ti for r in res for ti in r[0]]
				top_test_decoy_scores = [ti for r in res for ti in r[1]]
				clfs.extend([r[2] for r in res])
				all_test_target_scores.extend(top_test_target_scores)
				all_test_decoy_scores.extend(top_test_decoy_scores)

		logging.info("finished cross evals")
		final_classifier = ConsensusPredictor(clfs)
		
		d = {}
		d["tg_id"] = experiment.df.tg_num_id.values
		d["decoy"] = experiment.df.is_decoy.values
		for i in range(len(clfs)):
			s = clfs[i].score(experiment, True)
			experiment.set_and_rerank("classifier_score", s)
			d["clf%d_score" % i] = s.flatten()
			d["clf%d_rank1" % i] = experiment.df.is_top_peak.values

		for c in score_columns:
			d[c] = table[c]

		results, data_for_persistence = self.apply_classifier(final_classifier, experiment, test_exp,
															 all_test_target_scores,
															 all_test_decoy_scores, table)
		logging.info("calculated scoring and statistics")
		return  results, pd.DataFrame(d), data_for_persistence + (score_columns,)



	@profile
	def apply_loaded_scorer(self, table, loaded_scorer):

		final_classifier, mu, nu, df_raw_stat, loaded_score_columns = loaded_scorer

		prepared_table, __ = prepare_data_table(table, loaded_score_columns=loaded_score_columns)

		experiment = Experiment(prepared_table)

		final_score = final_classifier.score(experiment, True)
		experiment["d_score"] = (final_score - mu) / nu

		scored_table = self.enrich_table_with_results(table, experiment, df_raw_stat)

		return {'pyProph':(None, None, scored_table)}, None, None



	@profile
	def enrich_table_with_results(self, table, experiment, df_raw_stat):
		s_values, q_values = lookup_s_and_q_values_from_error_table(experiment["d_score"],
																	df_raw_stat)
		experiment["m_score"] = q_values
		experiment["s_value"] = s_values
		logging.info("mean m_score = %e, std_dev m_score = %e" % (np.mean(q_values),
					 np.std(q_values, ddof=1)))
		logging.info("mean s_value = %e, std_dev s_value = %e" % (np.mean(s_values),
					 np.std(s_values, ddof=1)))
		experiment.add_peak_group_rank()

		scored_table = table.join(experiment[["d_score", "m_score", "peak_group_rank"]])
		return scored_table



	@profile
	def apply_classifier(self, final_classifier, experiment, test_exp, all_test_target_scores,
						 all_test_decoy_scores, table):

		lambda_ = CONFIG.get("final_statistics.lambda")

		mu, nu, final_score = self.calculate_params_for_d_score(final_classifier, experiment)
		experiment["d_score"] = (final_score - mu) / nu

		all_tt_scores = experiment.get_top_target_peaks()["d_score"]
		
		def getRes(et):
			return (et.summary_table(), et.final_table(), et.enrich(table, experiment))
		
		is_test	 = CONFIG.get("is_test", False)
		
		if CONFIG.get("is_test", False):
			d = {
				'pyProph-flex':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									NormalNullModel(),
									PyProphFDRCalc(),
									PyProphStatCalc(),
									PyProphStatSampler()
							)), 
				'nonParam-flex':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									NonParamNullModel(),
									PyProphFDRCalc(),
									PyProphStatCalc(),
									PyProphStatSampler()
							)),
				'logNormal-flex':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									LogNormalNullModel(),
									PyProphFDRCalc(),
									PyProphStatCalc(),
									PyProphStatSampler()
							)),
				'nonParam-storey':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									NonParamNullModel(),
									StoreyFDRCalc(),
									PyProphStatCalc(),
									PyProphStatSampler()
							)),
				'nonParam-storey-jt':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									NonParamNullModel(),
									StoreyFDRCalc(),
									JTStatCalc(),
									PyProphStatSampler()
							))
				}
			
			d["res"] = d["nonParam-flex"]
		
			if test_exp is not None:
				muT, nuT, final_scoreT = self.calculate_params_for_d_score(final_classifier, test_exp)
				test_exp["d_score"] = (final_scoreT - muT) / nuT
				d['true_pyProph'] = getRes(FlexibleErrorTable(
						all_tt_scores, 
						test_exp.get_top_target_peaks()["d_score"],
						test_exp.get_top_decoy_peaks()["d_score"],
						lambda_,
						NormalNullModel(),
						PyProphFDRCalc(),
						PyProphStatCalc(),
						PyProphStatSampler()
					))
				d['true_nonParam'] = getRes(FlexibleErrorTable(
						all_tt_scores, 
						test_exp.get_top_target_peaks()["d_score"],
						test_exp.get_top_decoy_peaks()["d_score"],
						lambda_,
						NonParamNullModel(),
						PyProphFDRCalc(),
						PyProphStatCalc(),
						PyProphStatSampler()
					))
				d['true_logNormal'] = getRes(FlexibleErrorTable(
						all_tt_scores,
						test_exp.get_top_target_peaks()["d_score"],
						test_exp.get_top_decoy_peaks()["d_score"],
						lambda_,
						LogNormalNullModel(),
						PyProphFDRCalc(),
						PyProphStatCalc(),
						PyProphStatSampler()
					))
		else:
			 null_model 	= getNullModel(CONFIG.get("final_statistics.null_model"))
			 fdr_calc 		= getFDRCalc(CONFIG.get("final_statistics.fdr_calc"))
			 stat_calc 		= getStatCalc(CONFIG.get("final_statistics.stat_calc"))
			 stat_sampler 	= getStatSampler(CONFIG.get("final_statistics.stat_sampler"))
			 decoys_missing	= CONFIG.get("decoy.missing", 0.0)
			 d = dict(
			 		res = getRes(FlexibleErrorTable(
						all_tt_scores,
						all_test_target_scores,
						all_test_decoy_scores,
						lambda_,
						null_model,
						fdr_calc,
						stat_calc,
						stat_sampler,
						decoys_missing
					))
			 	)


		needed_to_persist = (final_classifier, mu, nu)
							 #error_table.df_raw_stat.loc[:, ["svalue", "qvalue", "cutoff"]])
		return d, needed_to_persist

	@profile
	def calculate_params_for_d_score(self, classifier, experiment):
		score = classifier.score(experiment, True)
		experiment.set_and_rerank("classifier_score", score)
		td_scores = experiment.get_top_decoy_peaks()["classifier_score"]
		
		mu = 0#mu = np.median(td_scores)
		nu = 1#nu = np.percentile(td_scores, 95.0)
				
		#mu, nu = mean_and_std_dev(td_scores)
		return mu, nu, score


@profile
def PyProphet():
	return HolyGostQuery(StandardSemiSupervisedTeacher(LDALearner()))
