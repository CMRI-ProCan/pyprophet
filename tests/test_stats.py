# encoding: utf-8
from __future__ import print_function

from pyprophet.stats import pi0est, qvalue, pemp, lfdr

import pandas as pd
pd.options.display.width = 220
pd.options.display.precision = 6

import numpy as np
import os
import shutil

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def test_0(tmpdir, regtest):
    os.chdir(tmpdir.strpath)
    data_path = os.path.join(DATA_FOLDER, "test_qvalue_ref_data.csv")
    shutil.copy(data_path, tmpdir.strpath)

    stat = pd.read_csv('test_qvalue_ref_data.csv', delimiter=',').sort_values("p")

    # For comparison with R/bioconductor reference implementation
    np.testing.assert_almost_equal(qvalue(stat['p'], 0.669926026474838, pfdr=False), stat['q_default'].values, decimal=4)
    np.testing.assert_almost_equal(qvalue(stat['p'], 0.669926026474838, pfdr=True), stat['q_pfdr'].values, decimal=4)

def test_1():
    stat = np.array([0, 1, 3, 2, 0.1, 0.5, 0.6, 0.3, 0.5, 0.6, 0.2, 0.5])
    stat0 = np.array([0.4, 0.2, 0.5, 1, 0.5, 0.7, 0.2, 0.4])

    np.testing.assert_almost_equal(pemp(stat, stat0),
                                   np.array([1.0, 0.125, 0.125, 0.125, 1.0, 0.25, 0.25, 0.75, 0.25, 0.25, 0.75, 0.25]))

def test_2(tmpdir, regtest):
    os.chdir(tmpdir.strpath)
    data_path = os.path.join(DATA_FOLDER, "test_lfdr_ref_data.csv")
    shutil.copy(data_path, tmpdir.strpath)

    stat = pd.read_csv('test_lfdr_ref_data.csv', delimiter=',').sort_values("p")

    # For comparison with R/bioconductor reference implementation
    np.testing.assert_almost_equal(lfdr(stat['p'], 0.669926026474838), stat['lfdr_default'].values, decimal=3)
    np.testing.assert_almost_equal(lfdr(stat['p'], 0.669926026474838, monotone = False), stat['lfdr_monotone_false'].values, decimal=3)
    np.testing.assert_almost_equal(lfdr(stat['p'], 0.669926026474838, transf="logit"), stat['lfdr_transf_logit'].values, decimal=1)
    np.testing.assert_almost_equal(lfdr(stat['p'], 0.669926026474838, eps=np.power(10.0,-2)), stat['lfdr_eps'].values, decimal=3)

    print(lfdr(stat['p'], 0.669926026474838), file=regtest)
    print(lfdr(stat['p'], 0.669926026474838, monotone = False), file=regtest)
    print(lfdr(stat['p'], 0.669926026474838, transf="logit"), file=regtest)
    print(lfdr(stat['p'], 0.669926026474838, eps=np.power(10.0,-2)), file=regtest)

def test_3(tmpdir):
    os.chdir(tmpdir.strpath)
    data_path = os.path.join(DATA_FOLDER, "test_lfdr_ref_data.csv")
    shutil.copy(data_path,tmpdir.strpath)

    stat = pd.read_csv('test_lfdr_ref_data.csv', delimiter=',').sort_values("p")

    # For comparison with R/bioconductor reference implementation
    # np.testing.assert_almost_equal(pi0est(stat['p'], lambda_ = 0.4)['pi0'], 0.6971609)
    # np.testing.assert_almost_equal(pi0est(stat['p'])['pi0'], 0.669926)
    # np.testing.assert_almost_equal(pi0est(stat['p'], lambda_ = np.arange(0.4,1.0,0.05), smooth_log_pi0 = True)['pi0'], 0.6718003)
    # np.testing.assert_almost_equal(pi0est(stat['p'], pi0_method = "bootstrap")['pi0'], 0.6763407)

    np.testing.assert_almost_equal(pi0est(stat['p'], lambda_ = 0.4)['pi0'], 0.697161)
    np.testing.assert_almost_equal(pi0est(stat['p'])['pi0'], 0.6685638)
    np.testing.assert_almost_equal(pi0est(stat['p'], lambda_ = np.arange(0.4,1.0,0.05), smooth_log_pi0 = True)['pi0'], 0.6658949)
    np.testing.assert_almost_equal(pi0est(stat['p'], pi0_method = "bootstrap")['pi0'], 0.6763406)


def test_random(regtest):
    np.random.seed(1)
    for i in (1, 2, 5, 10, 100):
        for j in (1, 2, 5, 10, 100):
            stat = np.random.random((i,))
            stat0 = np.random.random((j,))
            print(i, j, file=regtest)
            print(pemp(stat, stat0), file=regtest)
