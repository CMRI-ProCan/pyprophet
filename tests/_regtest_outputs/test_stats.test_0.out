without correction
     pvalue  percentile_positive  positive  negative        TP          FP        TN        FN  FDR       FNR  sens       FPR  qvalue  svalue
0  1.000000                 1.00     100.0       0.0  0.000000  100.000000  0.000000  0.000000  1.0       NaN   NaN  1.000000     1.0     NaN
1  1.000000                 1.00     100.0       0.0  0.000000  100.000000  0.000000  0.000000  1.0       NaN   NaN  1.000000     1.0     NaN
2  0.999998                 0.98      98.0       2.0 -1.999796   99.999794  0.000204  1.999796  1.0  0.999898   0.0  0.999998     1.0     1.0
3  0.999996                 0.97      97.0       3.0 -2.999592   99.999596  0.000408  2.999592  1.0  0.999864   0.0  0.999996     1.0     1.0
4  0.999994                 0.96      96.0       4.0 -3.999388   99.999390  0.000612  3.999388  1.0  0.999847   0.0  0.999994     1.0     1.0
      pvalue  percentile_positive  positive  negative        TP        FP          TN        FN  FDR       FNR  sens       FPR  qvalue  svalue
95  0.081633                 0.05       5.0      95.0 -3.163265  8.163265   91.836731  3.163265  1.0  0.033298   0.0  0.081633     1.0     1.0
96  0.061224                 0.04       4.0      96.0 -2.122449  6.122449   93.877548  2.122449  1.0  0.022109   0.0  0.061224     1.0     1.0
97  0.040816                 0.03       3.0      97.0 -1.081633  4.081633   95.918365  1.081633  1.0  0.011151   0.0  0.040816     1.0     1.0
98  0.020408                 0.02       2.0      98.0 -0.040816  2.040816   97.959183  0.040816  1.0  0.000416   0.0  0.020408     1.0     1.0
99  0.000000                 0.01       1.0      99.0  1.000000  0.000000  100.000000 -1.000000  0.0  0.000000   1.0  0.000000     0.0     1.0
with correction
     pvalue  percentile_positive  positive  negative        TP          FP        TN        FN  FDR  FNR  sens       FPR  qvalue  svalue
0  1.000000                 1.00     100.0       0.0  0.000000  100.000000  0.000000  0.000000  1.0  NaN   NaN  1.000000     1.0     NaN
1  1.000000                 1.00     100.0       0.0  0.000000  100.000000  0.000000  0.000000  1.0  NaN   NaN  1.000000     1.0     NaN
2  0.999998                 0.98      98.0       2.0 -1.999796   99.999794  0.000204  1.999796  1.0  1.0   0.0  0.999998     1.0     1.0
3  0.999996                 0.97      97.0       3.0 -2.999592   99.999596  0.000408  2.999592  1.0  1.0   0.0  0.999996     1.0     1.0
4  0.999994                 0.96      96.0       4.0 -3.999388   99.999390  0.000612  3.999388  1.0  1.0   0.0  0.999994     1.0     1.0
      pvalue  percentile_positive  positive  negative        TP        FP          TN        FN   FDR       FNR  sens       FPR  qvalue  svalue
95  0.081633                 0.05       5.0      95.0 -3.163265  8.163265   91.836731  3.163265  1.00  0.033298   0.0  0.081633    1.00     1.0
96  0.061224                 0.04       4.0      96.0 -2.122449  6.122449   93.877548  2.122449  1.00  0.022109   0.0  0.061224    1.00     1.0
97  0.040816                 0.03       3.0      97.0 -1.081633  4.081633   95.918365  1.081633  1.00  0.011151   0.0  0.040816    1.00     1.0
98  0.020408                 0.02       2.0      98.0 -0.040816  2.040816   97.959183  0.040816  1.00  0.000416   0.0  0.020408    1.00     1.0
99  0.000000                 0.01       1.0      99.0  1.000000  0.000000  100.000000 -1.000000  0.01  0.010000   1.0  0.000000    0.01     1.0
